src/cmd/services/m3comparator/main/querier.go

```diff
@@ -0,0 +1,221 @@
+	sync.Mutex
+}
+
+func noop() error { return nil }
+
+type seriesBlock []ts.Datapoint
+type tagMap map[string]string
+type series struct {
+	blocks []seriesBlock
+	tags   tagMap
+}
+
+func generateSeriesBlock(
+	start time.Time,
+	blockSize time.Duration,
+	resolution time.Duration,
+) seriesBlock {
+	numPoints := int(blockSize / resolution)
+	dps := make(seriesBlock, 0, numPoints)
+	for i := 0; i < numPoints; i++ {
+		dp := ts.Datapoint{
+			Timestamp: start.Add(resolution * time.Duration(i)),
+			Value:     rand.Float64(),
+		}
+
+		dps = append(dps, dp)
+	}
+
+	return dps
+}
+
+func generateSeries(
+	start time.Time,
+	end time.Time,
+	blockSize time.Duration,
+	resolution time.Duration,
+	tags tagMap,
+) (series, error) {
+	numBlocks := int(math.Ceil(float64(end.Sub(start)) / float64(blockSize)))
+	if numBlocks == 0 {
+		return series{}, fmt.Errorf("comparator querier: no blocks generated")
+	}
+
+	blocks := make([]seriesBlock, 0, numBlocks)
+	for i := 0; i < numBlocks; i++ {
+		blocks = append(blocks, generateSeriesBlock(start, blockSize, resolution))
+		start = start.Add(blockSize)
+	}
+
+	return series{
+		blocks: blocks,
+		tags:   tags,
+	}, nil
+}
+
+func (q *querier) generateOptions(
+	start time.Time,
+	blockSize time.Duration,
+) iteratorOptions {
+	return iteratorOptions{
+		start:         start,
+		blockSize:     blockSize,
+		encoderPool:   q.encoderPool,
+		iteratorPools: q.iteratorPools,
+	}
+}
+
+type seriesGen struct {
+	name string
+	res  time.Duration
+}
+
+// FetchCompressed fetches timeseries data based on a query.
+func (q *querier) FetchCompressed(
+	ctx context.Context,
+	query *storage.FetchQuery,
+	options *storage.FetchOptions,
+) (m3.SeriesFetchResult, m3.Cleanup, error) {
+	var (
+		defaultName = models.NewTagOptions().MetricName()
+
+		// TODO: take from config.
+		blockSize = time.Hour * 12
+		start     = query.Start.Truncate(blockSize)
+		end       = query.End.Truncate(blockSize).Add(blockSize)
+		opts      = q.generateOptions(start, blockSize)
+
+		// TODO: take from config.
+		gens = []seriesGen{
+			seriesGen{"foo", time.Second},
+			seriesGen{"bar", time.Second * 15},
+			seriesGen{"quail", time.Minute},
+		}
+
+		actualGens []seriesGen
+	)
+
+	q.Lock()
+	rand.Seed(start.Unix())
+	for _, matcher := range query.TagMatchers {
+		// filter if name, otherwise return all.
+		if bytes.Equal([]byte(defaultName), matcher.Name) {
+			value := string(matcher.Value)
+			for _, gen := range gens {
+				if value == gen.name {
+					actualGens = append(actualGens, gen)
+					break
+				}
+			}
+
+			break
+		}
+	}
+
+	if len(actualGens) == 0 {
+		actualGens = gens
+	}
+
+	seriesList := make([]series, 0, len(actualGens))
+	for _, gen := range actualGens {
+		tagMap := map[string]string{
+			"__name__": gen.name,
+			"foobar":   "qux",
+			"name":     gen.name,
+		}
+
+		series, err := generateSeries(start, end, blockSize, gen.res, tagMap)
+		if err != nil {
+			return m3.SeriesFetchResult{}, noop, err
```

Comment:
Hm, seems like we don't Unlock() before returning here? 

Can we just do `q.Lock(); defer q.Unlock()` when we take the lock?

Since primarily for test purposes doesn't seem worth avoiding missing the Unlock() so that `buildSeriesIterators` happens outside the lock?

Reply:
Yeah fair enough, don't think any implications in terms of results so should be good